### PR TITLE
drivers: i3c: npcx: introduce I3C driver

### PR DESCRIPTION
**Pull request by alvsun**
This implements basic driver to utilize the I3C IP block on NPCX.

1. I3C mode: Main controller mode only.
2. Transfer: Support SDR only.
3. IBI: Support Hot-Join, IBI(MDB). Controller request is not supported.
4. Support 3 I3C modules: I3C1(3.3V), I3C2(1.8V, espi mode), (I3C3 1.8V or 3.3V)

### CODE REVIEW

**Comment by alvsun** [comment]
This PR depends on #70932.
After #70932 is merged into ToT, we will begin the review.

**Comment by dcpleung** [comment]
Could you split this into separate commits? I can see the I3C driver change, clock control change, and SoC/board change.

**Comment by alvsun** [comment]
> Could you split this into separate commits? I can see the I3C driver change, clock control change, and SoC/board change.

Done

**Review by dcpleung — DISMISSED** [review]

File: drivers/i3c/i3c_npcx.c [hunk 1]
```diff
@@ -0,0 +1,2047 @@
...
+	}
+
+	if (config->common.dev_list.num_i3c == 0) {
+		/*
+		 * No i3c devices in dev tree. Just return so
+		 * we don't get errors doing cmds when there
+		 * are no devices listening/responding.
+		 */
+		return 0;
+	}
```
Comment (XenuIsWatching) [inline drivers/i3c/i3c_npcx.c:1224]: I'm not sure what the point of this is? I think someone want to see the nack on the i3c broadcast header if there are not i3c devices currently known in it's 'list.

I think you're doing this because of the `i3c_bus_init` in the driver init function so you will a return of `0` rather than an error. Would it make more sense to have it "not call" the `i3c_bus_init` if there is nothing in it's i3c device list instead?

File: drivers/i3c/i3c_npcx.c [hunk 2]
```diff
@@ -0,0 +1,2047 @@
...
+
+#ifdef CONFIG_I3C_NPCX_DMA
+	struct npcx_clk_cfg mdma_clk_subsys;
+	struct mdma_reg *mdma_base;
+#endif
+};
+
+struct npcx_i3c_data {
+	struct i3c_driver_data common;       /* Common i3c driver data */
+	struct k_sem lock_sem;               /* Mutex of i3c controller */
```
Comment (XenuIsWatching) [inline drivers/i3c/i3c_npcx.c:99]: shouldn't this be a 'mutex' to prevent a priority inversion

**Review by XenuIsWatching — CHANGES_REQUESTED** [review]
Looks Great! only a couple of questions I have here

File: drivers/i3c/i3c_npcx.c (hunk 2 above)
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done, thanks

File: drivers/i3c/i3c_npcx.c (hunk 1 above)
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done.
I have moved the target device checking before `i3c_bus_init` in `npcx_i3c_init`.

File: drivers/i3c/i3c_common.c
Comment (fabiobaltieri) [inline drivers/i3c/i3c_common.c:1]: this file should be in its own commit, usually a good idea to avoid mixing driver and subsystem changes unless the changes are related (like it would break the build to not submit them together)

File: drivers/i3c/i3c_npcx.c [hunk 3]
```diff
@@ -0,0 +1,2055 @@
...
+#include <zephyr/drivers/i3c.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/reset.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(npcx_i3c, CONFIG_I3C_LOG_LEVEL);
+
+#define NPCX_I3C_CHK_TIMEOUT 10000    /* Unit: us, timeout for checking register status */
+#define I3C_SCL_PP_FREQ_MAX  12500000 /* Unit: Mhz*/
+#define I3C_SCL_OD_FREQ_MAX  4170000  /* Unit: Mhz*/
```
Comment (fabiobaltieri) [inline drivers/i3c/i3c_npcx.c:26]: use the unit in the define naming, drop the comment, `I3C_SCL_OD_FREQ_MAX_MHZ` etc

File: drivers/i3c/i3c_npcx.c [hunk 4]
```diff
@@ -0,0 +1,2055 @@
...
+#define I3C_BUS_THigh_PP_MIN_NS 24  /* T_High period in push-pull mode */
+#define I3C_BUS_TLOW_OD_MIN_NS  200 /* T_LOW period in open-drain mode */
+
+#define PPBAUD_DIV_MAX (BIT(GET_FIELD_SZ(NPCX_I3C_MCONFIG_PPBAUD)) - 1) /* PPBAUD divider max */
+
+#define DAA_TGT_INFO_SZ 0x8 /* 8 bytes = PID(6) + BCR(1) + DCR(1) */
+#define BAMATCH_DIV     0x4 /* BAMATCH = APB4_CLK divided by four */
+
+/* Default maximum time we allow for an I3C transfer (unit:ms) */
+#define I3C_TRANS_TIMEOUT K_MSEC(100)
```
Comment (fabiobaltieri) [inline drivers/i3c/i3c_npcx.c:38]: I would spec this in MS (`I3C_TRANS_TIMEOUT_MS`) and use `K_MSEC` in the usage, just cor coherency with how other time units in the file are declared

File: drivers/i3c/i3c_npcx.c [hunk 5]
```diff
@@ -0,0 +1,2055 @@
...
+
+static int i3c_ctrl_wait_completion(const struct device *dev)
+{
+	struct npcx_i3c_data *const data = dev->data;
+
+	if (k_sem_take(&data->sync_sem, I3C_TRANS_TIMEOUT) == 0) {
+		return 0;
+	}
+
+	return -ETIMEDOUT;
```
Comment (fabiobaltieri) [inline drivers/i3c/i3c_npcx.c:158]: could you just `return k_sem_take(...)`?

File: drivers/i3c/i3c_npcx.c [hunk 6]
```diff
@@ -0,0 +1,2055 @@
...
+
+static inline void npcx_i3c_interrupt_enable(struct i3c_reg *inst, uint32_t mask)
+{
+	inst->MINTSET = mask;
+}
+
+static bool npcx_i3c_has_error(struct i3c_reg *inst)
+{
+	if (IS_BIT_SET(inst->MSTATUS, NPCX_I3C_MSTATUS_ERRWARN)) {
+
```
Comment (fabiobaltieri) [inline drivers/i3c/i3c_npcx.c:213]: nit: drop the empty line here

File: drivers/i3c/i3c_npcx.c [hunk 7]
```diff
@@ -0,0 +1,2055 @@
...
+		LOG_DBG("ERROR: MSTATUS 0x%08x MERRWARN 0x%08x", inst->MSTATUS, inst->MERRWARN);
```
Comment (fabiobaltieri) [inline drivers/i3c/i3c_npcx.c:214]: should this be a `LOG_ERR` or `LOG_WRN`?

File: drivers/i3c/i3c_npcx.c [hunk 8]
```diff
@@ -0,0 +1,2055 @@
...
+		mctrl |= BIT(NPCX_I3C_MCTRL_DIR);
+		SET_FIELD(mctrl, NPCX_I3C_MCTRL_RDTERM, read_sz); /* Set read length */
+	} else {
+		mctrl &= ~BIT(NPCX_I3C_MCTRL_DIR);
+	}
+
+	ret = npcx_i3c_send_request(inst, mctrl);
+	if (ret != 0) {
+		LOG_ERR("Request start error, %d", ret);
+	}
```
Comment (fabiobaltieri) [inline drivers/i3c/i3c_npcx.c:346]: return `err`? (and `return 0` at the end)

File: drivers/i3c/i3c_npcx.c [hunk 9]
```diff
@@ -0,0 +1,2055 @@
...
+ *
+ * return  0 success
+ *         -ECANCELED i3c state not as expected.
+ *         -ETIMEDOUT check MCTRLDONE timeout.
+ *         -ENOSYS    invalid use of request.
+ */
+static inline int npcx_i3c_request_emit_stop(struct i3c_reg *inst)
+{
+	uint32_t val = 0;
+	int ret = 0;
```
Comment (fabiobaltieri) [inline drivers/i3c/i3c_npcx.c:372]: no need to initialize `ret`, same in the other functions below, there's a few of these

File: drivers/i3c/i3c_npcx.c [hunk 10]
```diff
@@ -0,0 +1,2055 @@
...
+		ibi_rules |= BIT(NPCX_I3C_IBIRULES_NOBYTE);
+	}
+
+	/* Update the register */
+	inst->IBIRULES = ibi_rules;
+
+	LOG_DBG("MIBIRULES 0x%08x", ibi_rules);
+}
+
+int npcx_i3c_ibi_enable(const struct device *dev, struct i3c_device_desc *target)
```
Comment (fabiobaltieri) [inline drivers/i3c/i3c_npcx.c:1482]: `static`

File: drivers/i3c/i3c_npcx.c [hunk 11]
```diff
@@ -0,0 +1,2055 @@
...
+		 * enable controller to raise interrupt when a target
+		 * initiates IBI.
+		 */
+		inst->MINTSET = BIT(NPCX_I3C_MINTSET_TGTSTART);
+	}
+
+	return ret;
+}
+
+int npcx_i3c_ibi_disable(const struct device *dev, struct i3c_device_desc *target)
```
Comment (fabiobaltieri) [inline drivers/i3c/i3c_npcx.c:1592]: `static`

File: drivers/i3c/i3c_npcx.c [hunk 12]
```diff
@@ -0,0 +1,2055 @@
...
+	LOG_DBG("SCL_PP_FEQ MAX: %d", I3C_SCL_PP_FREQ_MAX);
+	LOG_DBG("SCL_OD_FEQ MAX: %d", I3C_SCL_OD_FREQ_MAX);
+	LOG_DBG("scl_pp: %d", scl_pp);
+	LOG_DBG("scl_od: %d", scl_od);
+	LOG_DBG("hdr: %d", ctrl_config->supported_hdr);
+
+	/* MCLKD = MCLK / I3C_DIV(1 or 2)
+	 * MCLKD must between 40 mhz to 50 mhz.
+	 */
+	if (mclkd == MHZ(45)) {
```
Comment (fabiobaltieri) [inline drivers/i3c/i3c_npcx.c:1785]: add a define for that 45MHZ (fancy, never realized we had that helper until now)

File: drivers/i3c/i3c_npcx.c [hunk 13]
```diff
@@ -0,0 +1,2055 @@
...
+			.mdma_clk_subsys = NPCX_DT_CLK_CFG_ITEM_BY_IDX(id, 2),                     \
+		))                                                                                 \
+		IF_ENABLED(CONFIG_I3C_NPCX_DMA, (                                                  \
+			.mdma_base = (struct mdma_reg *)DT_INST_REG_ADDR_BY_IDX(id, 1),            \
+		))                                                                                 \
+	};                                                                                         \
+	static struct npcx_i3c_data npcx_i3c_data_##id = {                                         \
+		.common.ctrl_config.scl.i3c = DT_INST_PROP_OR(id, i3c_scl_hz, 0),                  \
+		.common.ctrl_config.scl.i2c = DT_INST_PROP_OR(id, i2c_scl_hz, 0),                  \
+		.clocks.i3c_od_scl_hz = DT_INST_PROP_OR(id, i3c_od_scl_hz, 0),                     \
```
Comment (fabiobaltieri) [inline drivers/i3c/i3c_npcx.c:2043]: should this be in `struct npcx_i3c_config`? does not looks like you are changing it in runtime

if that's the case then you are left with the other two, I would consider saving those in the config structure as well and initializing the data copy at init time from the code, would save the flash imagine from having an entire copy of the data structure just to intiialize two fields at boot time and the rest with zero

**Comment by XenuIsWatching** [comment]
With https://github.com/zephyrproject-rtos/zephyr/pull/70773 now merged... you'll need to add the compact macro to https://github.com/zephyrproject-rtos/zephyr/blob/45205f865e18decbb9568c4691686dc679cbe961/drivers/i3c/i3c_shell.c#L88 and https://github.com/zephyrproject-rtos/zephyr/blob/45205f865e18decbb9568c4691686dc679cbe961/drivers/i3c/i3c_shell.c#L99 in a separate commit

File: drivers/i3c/i3c_common.c
Comment (alvsun) [reply drivers/i3c/i3c_common.c:1]: Done

File: drivers/i3c/i3c_npcx.c (hunk 3 above)
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done

File: drivers/i3c/i3c_npcx.c (hunk 7 above)
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done

File: drivers/i3c/i3c_npcx.c [hunk 14]
```diff
@@ -0,0 +1,2055 @@
...
+{
+	uint32_t val = 0;
+	int ret = 0;
+
+	SET_FIELD(val, NPCX_I3C_MCTRL_IBIRESP, MCTRL_IBIRESP_ACK);
+	SET_FIELD(val, NPCX_I3C_MCTRL_REQUEST, MCTRL_REQUEST_AUTOIBI);
+
+	ret = npcx_i3c_send_request(inst, val);
+	if (ret != 0) {
+		LOG_ERR("Request auto ibi error, %d", ret);
```
[the remark this answers is not in the document]
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c:303]: Done

File: drivers/i3c/i3c_npcx.c (hunk 8 above)
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c:353]: Done

File: drivers/i3c/i3c_npcx.c [hunk 15]
```diff
@@ -0,0 +1,2055 @@
...
+	}
+
+	return false;
+}
+
+static inline void npcx_i3c_status_clear_all(struct i3c_reg *inst)
+{
+	uint32_t mask = (BIT(NPCX_I3C_MSTATUS_TGTSTART) | BIT(NPCX_I3C_MSTATUS_MCTRLDONE) |
+			 BIT(NPCX_I3C_MSTATUS_COMPLETE) | BIT(NPCX_I3C_MSTATUS_IBIWON) |
+			 BIT(NPCX_I3C_MSTATUS_NOWCNTLR));
```
[the remark this answers is not in the document]
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done

File: drivers/i3c/i3c_npcx.c (hunk 9 above)
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done

File: drivers/i3c/i3c_npcx.c (hunk 10 above)
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done

**Comment by alvsun** [comment]
> With #70773 now merged... you'll need to add the compat macro to
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/45205f865e18decbb9568c4691686dc679cbe961/drivers/i3c/i3c_shell.c#L88
> 
> and
> https://github.com/zephyrproject-rtos/zephyr/blob/45205f865e18decbb9568c4691686dc679cbe961/drivers/i3c/i3c_shell.c#L99
> 
> in a separate commit

Done

File: drivers/i3c/i3c_npcx.c [hunk 16]
```diff
@@ -0,0 +1,2055 @@
...
+	config->irq_config_func(dev);
+
+	/* Initialize driver status machine */
+	set_oper_state(dev, NPCX_I3C_IDLE);
+
+	/* Check target device exist in device tree */
+	if (config->common.dev_list.num_i3c == 0) {
+		LOG_ERR("%s: No i3c target device in the device tree", __func__);
+		return -ENODEV;
+	}
```
[the remark this answers is not in the document]
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done

File: drivers/i3c/i3c_npcx.c (hunk 12 above)
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done

File: dts/arm/nuvoton/npcx/npcx4.dtsi [hunk 17]
```diff
@@ -14,6 +14,8 @@
 #include "npcx4/npcx4-espi-vws-map.dtsi"
 /* npcx4 series low-voltage io controls mapping table */
 #include "npcx4/npcx4-lvol-ctrl-map.dtsi"
+/* npcx4 series reset mapping table*/
```
Comment (fabiobaltieri) [inline dts/arm/nuvoton/npcx/npcx4.dtsi:17]: `/* npcx4 series reset mapping table */`

(missing space before the comment closing marker)

File: drivers/i3c/i3c_npcx.c [hunk 18]
```diff
@@ -0,0 +1,2053 @@
...
+	static struct npcx_i3c_data npcx_i3c_data_##id;                                            \
+	DEVICE_DT_INST_DEFINE(id, npcx_i3c_init, NULL, &npcx_i3c_data_##id, &npcx_i3c_config_##id, \
+			      POST_KERNEL, CONFIG_I3C_CONTROLLER_INIT_PRIORITY,                    \
+			      &npcx_i3c_driver_api);                                               \
+	static void npcx_i3c_config_func_##id(const struct device *dev)                            \
+	{                                                                                          \
+		IRQ_CONNECT(DT_INST_IRQN(id), DT_INST_IRQ(id, priority), npcx_i3c_isr,             \
+			    DEVICE_DT_INST_GET(id), 0);                                            \
+		irq_enable(DT_INST_IRQN(id));                                                      \
+	};
```
Comment (fabiobaltieri) [inline drivers/i3c/i3c_npcx.c:2051]: you could move this to the top of the macro, then you won't have to forward declare it

Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done

File: dts/arm/nuvoton/npcx/npcx4.dtsi (hunk 17 above)
Comment (alvsun) [reply dts/arm/nuvoton/npcx/npcx4.dtsi]: Done

File: drivers/i3c/i3c_shell.c [hunk 19]
```diff
@@ -86,6 +86,7 @@ struct i3c_ctrl {
 
 DT_FOREACH_STATUS_OKAY(cdns_i3c, I3C_CTRL_FN)
 DT_FOREACH_STATUS_OKAY(nxp_mcux_i3c, I3C_CTRL_FN)
+DT_FOREACH_STATUS_OKAY(nuvoton_npcx_i3c, I3C_CTRL_FN)
```
Comment (XenuIsWatching) [inline drivers/i3c/i3c_shell.c:89]: nit: This should be in alphabetical order

Comment (alvsun) [reply drivers/i3c/i3c_shell.c:88]: Done

**Review by XenuIsWatching — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Review by MulinChao — APPROVED** [review]
LGTM.

**Review by ChiHuaL — APPROVED** [review]

File: drivers/i3c/i3c_npcx.c [hunk 20]
```diff
@@ -0,0 +1,2052 @@
...
+				i3c_ccc_is_payload_broadcast(payload) ? "broadcast" : "direct",
+				ret);
+			ret = xfered_len;
+
+			goto out_do_ccc;
+		}
+
+		/* Write back the transferred bytes */
+		payload->ccc.num_xfer = xfered_len;
+	}
```
Comment (XenuIsWatching) [inline drivers/i3c/i3c_npcx.c:1282]: How does this distinguish the difference between the defining byte and payload for the direct CCC in I3C v1.1?

Comment (alvsun) [reply drivers/i3c/i3c_npcx.c:1282]: Only the i3c_ccc APIs will know the difference (e.g., i3c_ccc_do_getstatus()). 
This part handles both broadcast and direct CCC write data operations in the earlier stage(marked as red in the picture below).

image
